streams/src/main/java/org/apache/kafka/streams/state/internals/ChangeLoggingKeyValueBytesStore.java

```diff
@@ -68,9 +68,10 @@ public void put(final Bytes key,
     @Override
     public byte[] putIfAbsent(final Bytes key,
                               final byte[] value) {
-        final byte[] previous = get(key);
+        final byte[] previous = inner.putIfAbsent(key, value);
         if (previous == null) {
-            put(key, value);
+            // then it was absent
+            changeLogger.logChange(key, value);
```

Comment:
This change would make it impossible for underlying stores to implement an atomic update. At the least, it prohibits a performance optimization if the underlying store supports `putIfAbsent` natively.

It could cause concurrency bugs, if the underlying store were attempting to provide thread-safe access via IQ. I'm not too sure about this, though.

Reply:
I cannot follow.

Reply:
The prior logic splits one call (`putIfAbsent`) into two separate operations (`get`, followed by `put`). The underlying store has its own `putIfAbsent` method, which we would never call through to.

Reply:
It's unexpected for a "pass-though" layer to actually change the behavior and semantics of an operation.

Reply:
Well, it's single threaded so it does not really matter too much. I understand that it is a change, but you mentioned 

> It could cause concurrency bugs, if the underlying store were attempting to provide thread-safe access via IQ. I'm not too sure about this, though.

that why I was asking. This was not an issue in existing code -- I guess the code was written this way to "share" the part that does the actually `changelog.logChange()` call

Reply:
Thus, I agree that the change makes sense -- it was just your comment that confused me.

Reply:
Yes, I agree. It also looked to me like the intent was to share the logging portion of the `put` call. Normally, this would be ideal, but in this case it seems to violate the expectations of store implementers.

It is true that Streams's processing is single-threaded. I should have been more explicit about the nature of the concurrency behavior I had in mind... Specifically, I was thinking about concurrent read access via IQ, but in retrospect, since IQ users don't have any visibility into the exact state of record processing, the concurrent read story should be identical whether Streams performs a single atomic operation, or two separate ones.

Therefore, I guess the concurrency thing is a non-issue. Sorry for the confusion.

Reply:
All good :)